File: output/GPT3.5-Turbo/xCodeEval_4000/C++/C/10fdacb4ed7655f969fdcac95872007f.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define N 200100

typedef struct {
    int be, en;
    int maxv, add, sum;
} segment_node;

typedef struct {
    int l;
    segment_node tree[N * 4];
} segment_tree;

typedef struct {
    int l, r, maxv, deep;
    double p;
} q;

segment_tree sgt;
q que[5010];
int son[5010][5010];
int arr[100100], fa[5010];
double dp[5010][5010];
int n, nq, be, en;

int max1(int a, int b) {
    return b < a ? a : b;
}

void build(segment_tree *sgt, int no, int l, int r, int orig, int *a) {
    if (l > r) r = l;
    if (l == r) {
        sgt->tree[no].be = sgt->tree[no].en = l;
        sgt->tree[no].add = 0;
        if (a == NULL) sgt->tree[no].maxv = orig;
        else sgt->tree[no].maxv = a[l];
        sgt->tree[no].add = 0;
        sgt->tree[no].sum = sgt->tree[no].maxv;
        return;
    }
    sgt->tree[no].be = l; sgt->tree[no].en = r;
    int mid = (l + r) / 2;
    build(sgt, no * 2, l, mid, orig, a);
    build(sgt, no * 2 + 1, mid + 1, r, orig, a);
    sgt->tree[no].sum = sgt->tree[no * 2].sum + sgt->tree[no * 2 + 1].sum;
    sgt->tree[no].add = 0;
    sgt->tree[no].maxv = max1(sgt->tree[no * 2].maxv, sgt->tree[no * 2 + 1].maxv);
}

void relax(segment_tree *sgt, int no) {
    int le = no * 2, ri = no * 2 + 1;
    sgt->tree[le].add += sgt->tree[no].add;
    sgt->tree[le].sum += sgt->tree[no].add * (sgt->tree[le].en - sgt->tree[le].be + 1);
    sgt->tree[le].maxv += sgt->tree[no].add;
    sgt->tree[ri].add += sgt->tree[no].add;
    sgt->tree[ri].sum += sgt->tree[no].add * (sgt->tree[ri].en - sgt->tree[ri].be + 1);
    sgt->tree[ri].maxv += sgt->tree[no].add;
    sgt->tree[no].add = 0;
}

void down(segment_tree *sgt, int l, int r, int no, int ranadd) {
    if (l <= sgt->tree[no].be && r >= sgt->tree[no].en) {
        sgt->tree[no].add += ranadd;
        sgt->tree[no].sum += ranadd * (sgt->tree[no].en - sgt->tree[no].be + 1);
        sgt->tree[no].maxv += ranadd;
        return;
    }
    if (sgt->tree[no].add && sgt->tree[no].be != sgt->tree[no].en) relax(sgt, no);
    int mid = (sgt->tree[no].be + sgt->tree[no].en) >> 1;
    if (r >= sgt->tree[no].be && l <= mid) down(sgt, l, r, no * 2, ranadd);
    if (r >= mid + 1 && l <= sgt->tree[no].en) down(sgt, l, r, no * 2 + 1, ranadd);
    sgt->tree[no].sum = sgt->tree[no * 2].sum + sgt->tree[no * 2 + 1].sum;
    sgt->tree[no].maxv = max1(sgt->tree[no * 2].maxv, sgt->tree[no * 2 + 1].maxv);
}

int getmax(segment_tree *sgt, int l, int r, int no) {
    if (l > r) return 0;
    if (l <= sgt->tree[no].be && r >= sgt->tree[no].en)
        return sgt->tree[no].maxv;
    if (sgt->tree[no].add && sgt->tree[no].be != sgt->tree[no].en) relax(sgt, no);
    int ans = 0;
    int mid = (sgt->tree[no].be + sgt->tree[no].en) >> 1;
    if (r >= sgt->tree[no].be && l <= mid) ans = max1(ans, getmax(sgt, l, r, no * 2));
    if (r >= mid + 1 && l <= sgt->tree[no].en) ans = max1(ans, getmax(sgt, l, r, no * 2 + 1));
    return ans;
}

void dfs(int no, int deep) {
    que[no].maxv = getmax(&sgt, que[no].l, que[no].r, 1);
    be = max1(be, que[no].maxv);
    en = max1(en, que[no].maxv + deep + 1);
    que[no].deep = deep;
    for (int i = 0; i < son[no][0]; i++)
        dfs(son[no][i], deep + 1);
}

void dfsdp(int no) {
    dp[no][0] = 1;
    for (int i = 0; i < son[no][0]; i++) {
        int rno = son[no][i];
        dfsdp(rno);
        double sum1 = dp[no][0], sum2 = dp[rno][0], p = que[rno].p;
        for (int link1 = 0, link2 = 0; link1 < dp[no][0] || link2 < dp[rno][0]; ) {
            dp[no][link1 + link2] = sum1 * sum2 * (1 - p);
            if (link1 == dp[no][0] - 1) {
                link2++;
                if (link2 == dp[rno][0]) break;
                sum2 += dp[rno][link2];
            } else if (link2 == dp[rno][0] - 1) {
                link1++;
                sum1 += dp[no][link1];
            } else if (dp[no][link1 + 1] > dp[rno][link2 + 1]) {
                link2++;
                sum2 += dp[rno][link2];
            } else {
                link1++;
                sum1 += dp[no][link1];
            }
        }
        for (int i = dp[no][0]; i > 0; i--)
            dp[no][i] -= dp[no][i - 1];
        int rsz = 1;
        for (int i = 1; i < dp[no][0]; i++) {
            if (dp[no][i] == dp[no][i - 1])
                dp[no][rsz - 1] += dp[no][i];
            else dp[no][rsz++] = dp[no][i];
        }
        dp[no][0] = rsz;
        sum1 = dp[no][0], sum2 = dp[rno][0];
        for (int link1 = 0, link2 = 0; link1 < dp[no][0] || link2 < dp[rno][0]; ) {
            if (sum1 * sum2 > 1e-10)
                dp[no][link1 + link2] = sum1 * sum2 * p;
            if (link1 == dp[no][0] - 1) {
                link2++;
                if (link2 == dp[rno][0]) break;
                sum2 += dp[rno][link2];
            } else if (link2 == dp[rno][0] - 1) {
                link1++;
                sum1 += dp[no][link1];
            } else if (dp[no][link1 + 1] > dp[rno][link2 + 1]) {
                link2++;
                sum2 += dp[rno][link2];
            } else {
                link1++;
                sum1 += dp[no][link1];
            }
        }
        for (int i = dp[no][0]; i > 0; i--)
            dp[no][i] -= dp[no][i - 1];
        rsz = 1;
        for (int i = 1; i < dp[no][0]; i++) {
            if (dp[no][i] == dp[no][i - 1])
                dp[no][rsz - 1] += dp[no][i];
            else dp[no][rsz++] = dp[no][i];
        }
        dp[no][0] = rsz;
        int rsz = 1;
        for (int link1 = 0, link2 = 0; link1 < dp[no][0] || link2 < dp[rno][0]; ) {
            if (link2 == dp[rno][0]) {
                dp[no][rsz++] = dp[no][link1++];
            } else if (link1 == dp[no][0]) {
                dp[no][rsz++] = dp[rno][link2++];
            } else if (dp[no][link1] > dp[rno][link2]) {
                dp[no][rsz++] = dp[rno][link2++];
            } else if (dp[no][link1] == dp[rno][link2]) {
                dp[no][rsz++] = dp[no][link1++] + dp[rno][link2++];
            } else if (dp[no][link1] < dp[rno][link2]) {
                dp[no][rsz++] = dp[no][link1++];
            }
        }
        dp[no][0] = rsz;
    }
}

int main() {
    scanf("%d%d", &n, &nq);
    for (int i = 0; i < n; i++) scanf("%d", &arr[i]);
    for (int i = 0; i < nq; i++) {
        scanf("%d%d%lf", &que[i].l, &que[i].r, &que[i].p);
        que[i].l--;
        que[i].r--;
    }
    memset(fa, -1, sizeof(fa));
    for (int i = 0; i < nq; i++) {
        for (int j = 0; j < i; j++) {
            if (j != i && que[j].r - que[j].l == que[i].r - que[i].l && que[j].r == que[i].r)
                fa[i] = j;
        }
        if (fa[i] != -1) continue;
        for (int j = 0; j < nq; j++) {
            if (j != i && que[j].r - que[j].l == que[i].r - que[i].l)
                continue;
            if (que[j].l <= que[i].l && que[j].r >= que[i].r) {
                if (fa[i] == -1) {
                    fa[i] = j;
                    continue;
                }
                if (que[j].r - que[j].l <= que[fa[i]].r - que[fa[i]].l) {
                    fa[i] = j;
                    continue;
                }
            }
        }
    }
    for (int i = 0; i < nq; i++) {
        if (fa[i] != -1) son[fa[i]][son[fa[i]][0]++] = i;
        else {
            fa[i] = nq;
            son[nq][son[nq][0]++] = i;
        }
    }
    que[nq].l = 0; que[nq].r = n - 1;
    build(&sgt, 1, 0, n - 1, 0, arr);

    dfs(nq, 0);
    dfsdp(nq);
    double ans = 0;
    for (int i = 0; i < dp[nq][0]; i++) {
        ans += dp[nq][i] * dp[nq][i];
    }
    printf("%.10lf\n", ans);

    return 0;
}

File: output/GPT3.5-Turbo/xCodeEval_4000/C++/Java/10fdacb4ed7655f969fdcac95872007f.java
import java.util.*;
import java.io.*;

class Main {
    static class SegmentNode<T> {
        int be, en;
        T maxv, add, sum;
    }

    static class SegmentTree<T> {
        int l;
        SegmentNode<T>[] tree;

        public SegmentTree(int n) {
            l = 0;
            tree = new SegmentNode[n * 4];
            for (int i = 0; i < n * 4; i++) {
                tree[i] = new SegmentNode<>();
            }
        }

        private int gleft(int no) {
            return no << 1;
        }

        private int gright(int no) {
            return (no << 1) + 1;
        }

        private int gfa(int no) {
            return no >> 1;
        }

        public void build(int no, int l, int r, T orig, T[] a) {
            if (l > r) {
                r = l;
            }
            if (l == r) {
                tree[no].be = tree[no].en = l;
                tree[no].add = null;
                if (a == null) {
                    tree[no].maxv = orig;
                } else {
                    tree[no].maxv = a[l];
                }
                tree[no].add = null;
                tree[no].sum = tree[no].maxv;
                return;
            }
            tree[no].be = l;
            tree[no].en = r;
            int mid = (l + r) / 2;
            build(gleft(no), l, mid, orig, a);
            build(gright(no), mid + 1, r, orig, a);
            tree[no].sum = tree[gleft(no)].sum + tree[gright(no)].sum;
            tree[no].add = null;
            tree[no].maxv = max1(tree[gleft(no)].maxv, tree[gright(no)].maxv);
        }

        private void relax(int no) {
            int le = gleft(no), ri = gright(no);
            tree[le].add = add(tree[le].add, tree[no].add);
            tree[le].sum = add(tree[le].sum, multiply(tree[no].add, tree[le].en - tree[le].be + 1));
            tree[le].maxv = add(tree[le].maxv, tree[no].add);
            tree[ri].add = add(tree[ri].add, tree[no].add);
            tree[ri].sum = add(tree[ri].sum, multiply(tree[no].add, tree[ri].en - tree[ri].be + 1));
            tree[ri].maxv = add(tree[ri].maxv, tree[no].add);
            tree[no].add = null;
        }

        private void down(int l, int r, int no, T ranadd) {
            if (l <= tree[no].be && r >= tree[no].en) {
                tree[no].add = add(tree[no].add, ranadd);
                tree[no].sum = add(tree[no].sum, multiply(ranadd, tree[no].en - tree[no].be + 1));
                tree[no].maxv = add(tree[no].maxv, ranadd);
                return;
            }
            if (tree[no].add != null && tree[no].be != tree[no].en) {
                relax(no);
            }
            int mid = (tree[no].be + tree[no].en) >> 1;
            if (r >= tree[no].be && l <= mid) {
                down(l, r, gleft(no), ranadd);
            }
            if (r >= mid + 1 && l <= tree[no].en) {
                down(l, r, gright(no), ranadd);
            }
            tree[no].sum = add(tree[gleft(no)].sum, tree[gright(no)].sum);
            tree[no].maxv = max1(tree[gleft(no)].maxv, tree[gright(no)].maxv);
        }

        public T getmax(int l, int r, int no) {
            if (l > r) {
                return null;
            }
            if (l <= tree[no].be && r >= tree[no].en) {
                return tree[no].maxv;
            }
            if (tree[no].add != null && tree[no].be != tree[no].en) {
                relax(no);
            }
            T ans = null;
            int mid = (tree[no].be + tree[no].en) >> 1;
            if (r >= tree[no].be && l <= mid) {
                ans = max1(ans, getmax(l, r, gleft(no)));
            }
            if (r >= mid + 1 && l <= tree[no].en) {
                ans = max1(ans, getmax(l, r, gright(no)));
            }
            return ans;
        }

        private T add(T a, T b) {
            // Add implementation here
            return null;
        }

        private T multiply(T a, int b) {
            // Multiply implementation here
            return null;
        }

        private T max1(T a, T b) {
            // Max implementation here
            return null;
        }
    }

    static class Query {
        int l, r, maxv, deep;
        double p;

        public Query(int l, int r, double p) {
            this.l = l;
            this.r = r;
            this.p = p;
        }
    }

    static int be, en;
    static Query[] que;
    static List<Integer>[] son;
    static int[] arr, fa;
    static List<Pair<Double, Integer>>[] dp;
    static int n, nq;

    static void dfs(int no, int deep) {
        que[no].maxv = sgt.getmax(que[no].l, que[no].r, 1);
        be = Math.max(be, que[no].maxv);
        en = Math.max(en, que[no].maxv + deep + 1);
        que[no].deep = deep;
        for (int i = 0; i < son[no].size(); i++) {
            dfs(son[no].get(i), deep + 1);
        }
    }

    static Pair<Double, Integer> add(Pair<Double, Integer> a, Pair<Double, Integer> b) {
        return new Pair<>(a.first + b.first, a.second);
    }

    static void dfsdp(int no) {
        dp[no].add(new Pair<>(1.0, que[no].maxv));
        for (int i = 0; i < son[no].size(); i++) {
            int rno = son[no].get(i);
            dfsdp(rno);
            List<Pair<Double, Integer>> rdp = new ArrayList<>();
            List<Pair<Double, Integer>> rdp1 = new ArrayList<>();
            double sum1 = dp[no].get(0).first, sum2 = dp[rno].get(0).first, p = que[rno].p;
            for (int link1 = 0, link2 = 0; link1 < dp[no].size() || link2 < dp[rno].size(); ) {
                rdp.add(new Pair<>(sum1 * sum2 * (1 - p), Math.max(dp[no].get(link1).second, dp[rno].get(link2).second)));
                if (link1 == dp[no].size() - 1) {
                    link2++;
                    if (link2 == dp[rno].size()) {
                        break;
                    }
                    sum2 += dp[rno].get(link2).first;
                } else if (link2 == dp[rno].size() - 1) {
                    link1++;
                    sum1 += dp[no].get(link1).first;
                } else if (dp[no].get(link1 + 1).second > dp[rno].get(link2 + 1).second) {
                    link2++;
                    sum2 += dp[rno].get(link2).first;
                } else {
                    link1++;
                    sum1 += dp[no].get(link1).first;
                }
            }
            for (int j = rdp.size() - 1; j > 0; j--) {
                rdp.get(j).first -= rdp.get(j - 1).first;
            }
            int rsz = 1;
            for (int j = 1; j < rdp.size(); j++) {
                if (rdp.get(j).second == rdp.get(j - 1).second) {
                    rdp.get(rsz - 1).first += rdp.get(j).first;
                } else {
                    rdp.set(rsz++, rdp.get(j));
                }
            }
            rdp.subList(rsz, rdp.size()).clear();

            sum1 = dp[no].get(0).first;
            sum2 = dp[rno].get(0).first;
            for (int link1 = 0, link2 = 0; link1 < dp[no].size() || link2 < dp[rno].size(); ) {
                if (sum1 * sum2 > 1e-10) {
                    rdp1.add(new Pair<>(sum1 * sum2 * p, Math.max(dp[no].get(link1).second, dp[rno].get(link2).second + 1)));
                }
                if (link1 == dp[no].size() - 1) {
                    link2++;
                    if (link2 == dp[rno].size()) {
                        break;
                    }
                    sum2 += dp[rno].get(link2).first;
                } else if (link2 == dp[rno].size() - 1) {
                    link1++;
                    sum1 += dp[no].get(link1).first;
                } else if (dp[no].get(link1 + 1).second > dp[rno].get(link2 + 1).second + 1) {
                    link2++;
                    sum2 += dp[rno].get(link2).first;
                } else {
                    link1++;
                    sum1 += dp[no].get(link1).first;
                }
            }
            for (int j = rdp1.size() - 1; j > 0; j--) {
                rdp1.get(j).first -= rdp1.get(j - 1).first;
            }
            rsz = 1;
            for (int j = 1; j < rdp1.size(); j++) {
                if (rdp1.get(j).second == rdp1.get(j - 1).second) {
                    rdp1.get(rsz - 1).first += rdp1.get(j).first;
                } else {
                    rdp1.set(rsz++, rdp1.get(j));
                }
            }
            rdp1.subList(rsz, rdp1.size()).clear();

            dp[no].clear();
            for (int link1 = 0, link2 = 0; link1 < rdp.size() || link2 < rdp1.size(); ) {
                if (link2 == rdp1.size()) {
                    dp[no].add(rdp.get(link1));
                    link1++;
                } else if (link1 == rdp.size()) {
                    dp[no].add(rdp1.get(link2));
                    link2++;
                } else if (rdp.get(link1).second > rdp1.get(link2).second) {
                    dp[no].add(rdp1.get(link2));
                    link2++;
                } else if (rdp.get(link1).second == rdp1.get(link2).second) {
                    dp[no].add(add(rdp.get(link1), rdp1.get(link2)));
                    link2++;
                    link1++;
                } else if (rdp.get(link1).second < rdp1.get(link2).second) {
                    dp[no].add(rdp.get(link1));
                    link1++;
                }
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] line = br.readLine().split(" ");
        n = Integer.parseInt(line[0]);
        nq = Integer.parseInt(line[1]);
        arr = new int[n];
        que = new Query[nq + 1];
        son = new List[nq + 1];
        dp = new List[nq + 1];
        fa = new int[nq];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        for (int i = 0; i <= nq; i++) {
            son[i] = new ArrayList<>();
            dp[i] = new ArrayList<>();
        }
        for (int i = 0; i < nq; i++) {
            line = br.readLine().split(" ");
            int l = Integer.parseInt(line[0]) - 1;
            int r = Integer.parseInt(line[1]) - 1;
            double p = Double.parseDouble(line[2]);
            que[i] = new Query(l, r, p);
        }
        Arrays.fill(fa, -1);
        for (int i = 0; i < nq; i++) {
            for (int j = 0; j < i; j++) {
                if (j != i && que[j].r - que[j].l == que[i].r - que[i].l && que[j].r == que[i].r) {
                    fa[i] = j;
                }
            }
            if (fa[i] != -1) {
                continue;
            }
            for (int j = 0; j < nq; j++) {
                if (j != i) {
                    if (que[j].r - que[j].l == que[i].r - que[i].l) {
                        continue;
                    }
                    if (que[j].l <= que[i].l && que[j].r >= que[i].r) {
                        if (fa[i] == -1) {
                            fa[i] = j;
                            continue;
                        }
                        if (que[j].r - que[j].l <= que[fa[i]].r - que[fa[i]].l) {
                            fa[i] = j;
                            continue;
                        }
                    }
                }
            }
        }
        for (int i = 0; i < nq; i++) {
            if (fa[i] != -1) {
                son[fa[i]].add(i);
            } else {
                fa[i] = nq;
                son[nq].add(i);
            }
        }
        que[nq] = new Query(0, n - 1, 0);
        SegmentTree<Integer> sgt = new SegmentTree<>(n);
        sgt.build(1, 0, n - 1, 0, arr);
        dfs(nq, 0);
        dfsdp(nq);
        double ans = 0;
        for (int i = 0; i < dp[nq].size(); i++) {
            ans += dp[nq].get(i).first * dp[nq].get(i).second;
        }
        System.out.printf("%.10f\n", ans);
    }
}
